# Task 2

## General instructions

First, update the following cell to have a link to your Homework 1 GitHub repository

**UPDATE THIS LINK**
https://github.com/maxwellpatt/eds220-hwk-2

Add comments for all your code and commit as needed. Err on the side of commenting and commiting too much for now.


## About the data

In this task you will use [Air Quality Index (AQI)](https://www.airnow.gov/aqi/aqi-basics/) data from the [US Environmental Protection Agency](https://www.epa.gov) to visualize the impact on the AQI of the 2017 [Thomas Fire](https://en.wikipedia.org/wiki/Thomas_Fire) in Santa Barbara County. 

## FINAL CODE

You will use the next cell at the end of the task. Leave it blank for now. 

In [1]:
### ===== FINAL CODE =====

## 1

1. Go to [EPA's website on Air Quality Data Collected at Outdoor Monitors Across the US](https://www.epa.gov/outdoor-air-quality-data). 
2. Under "Donwload Data" click on "Pre-generated Data Files"
3. Click on "Tables of Daily AQI"
4. Copy the URL to the 2017 Daily AQI **by County** zip file `daily_aqi_by_county_2017.zip`
5. In the next code cell read in the data from the URL using the `pd.read_csv` function and store it as `aqi_17`. 
6. In the same cell, read in the data for the 2018 Daily AQI by County zip file and store it as `aqi_18`.

In [4]:
# read in data
# import pandas library
import pandas as pd

# read in data
aqi_17 = pd.read_csv("https://aqs.epa.gov/aqsweb/airdata/daily_aqi_by_county_2017.zip")
aqi_18 = pd.read_csv("https://aqs.epa.gov/aqsweb/airdata/daily_aqi_by_county_2018.zip")

## 2 
(a) and (b) Use the next two cells to look at the head of both data frames.

(c) Use this cell to make some other preliminary data exploration of your choosing.

In [6]:
# show head of aqi_17
aqi_17.head

<bound method NDFrame.head of        State Name county Name  State Code  County Code        Date  AQI  \
0         Alabama     Baldwin           1            3  2017-01-01   21   
1         Alabama     Baldwin           1            3  2017-01-04   22   
2         Alabama     Baldwin           1            3  2017-01-10   19   
3         Alabama     Baldwin           1            3  2017-01-13   30   
4         Alabama     Baldwin           1            3  2017-01-16   16   
...           ...         ...         ...          ...         ...  ...   
326796    Wyoming      Weston          56           45  2017-12-27   39   
326797    Wyoming      Weston          56           45  2017-12-28   36   
326798    Wyoming      Weston          56           45  2017-12-29   31   
326799    Wyoming      Weston          56           45  2017-12-30   31   
326800    Wyoming      Weston          56           45  2017-12-31   26   

       Category Defining Parameter Defining Site  Number of Sites Rep

In [8]:
# show head of aqi_18
aqi_18.head

<bound method NDFrame.head of        State Name county Name  State Code  County Code        Date  AQI  \
0         Alabama     Baldwin           1            3  2018-01-02   32   
1         Alabama     Baldwin           1            3  2018-01-05   34   
2         Alabama     Baldwin           1            3  2018-01-08   15   
3         Alabama     Baldwin           1            3  2018-01-11   19   
4         Alabama     Baldwin           1            3  2018-01-14   25   
...           ...         ...         ...          ...         ...  ...   
327532    Wyoming      Weston          56           45  2018-12-27   36   
327533    Wyoming      Weston          56           45  2018-12-28   35   
327534    Wyoming      Weston          56           45  2018-12-29   35   
327535    Wyoming      Weston          56           45  2018-12-30   31   
327536    Wyoming      Weston          56           45  2018-12-31   35   

       Category Defining Parameter Defining Site  Number of Sites Rep

In [13]:
# get datatypes of each column for each data set
dtypes_17 = aqi_17.dtypes
dtypes_18 = aqi_18.dtypes

# show results
print(dtypes_17)
print(dtypes_18)

# store info of each data set
info_17 = aqi_17.info
info_18 = aqi_18.info

# show results
print(info_17)
print(info_18)

State Name                   object
county Name                  object
State Code                    int64
County Code                   int64
Date                         object
AQI                           int64
Category                     object
Defining Parameter           object
Defining Site                object
Number of Sites Reporting     int64
dtype: object
State Name                   object
county Name                  object
State Code                    int64
County Code                   int64
Date                         object
AQI                           int64
Category                     object
Defining Parameter           object
Defining Site                object
Number of Sites Reporting     int64
dtype: object


## 3
We currently have two separate dataframes. For this exercise we will need to "glue" them one on top of the other. The pandas function `pd.concat()` can achieve this. 

Pass `[aqi_17, aqi_18]` as the input of `pd.concat()` and store the output as  `aqi`.  
In the next line run `aqi`.

👀 NOTE: When we concatenate like this, without any extra parameters for `pd.concat()` the indices for the two dataframes are just "glued together", the index of the resulting dataframe is not updated to start from 0. Notice the index of `aqi` ends in 327536 while it has 654338 rows.


In [ ]:
# solve exercise here

## 4

Run the follwoing cell and read the comments to understand how the column names are being updated.

In [ ]:
# initial column names: notice caps and spaces (difficult to work with!)
print(aqi.columns, '\n')

# re-assign the column names - .str.lower() makes them lower case
aqi.columns = aqi.columns.str.lower()
print(aqi.columns, '\n')

#  re-assign the column names again - .str.replace(' ','_') replaces the space for _
aqi.columns = aqi.columns.str.replace(' ','_')
print(aqi.columns)

# as a "one liner" you could achieve this column name cleaning like this:
# aqi.columns = aqi.columns.str.lower().str.replace(' ','_')

## 5 
In the next cell:
- Select only data from `Santa Barbara` county and store in a new variable `aqi_sb`.
- Remove the `state_name`, `county_name`, `state_code` and `county_code` columns from `aqi_sb`. 
- Use the `dtypes` attribute to check the data types of the columns. What do you notice for the date column?

In [ ]:
# solve exercise here

## 6
In the next cell:
1. Update the date column of `aqi_sb` to be a datetime object.
2. Update the index of `aqi_sb` to be the date column.

In [ ]:
# solve exercise here

## 7
In the next cell, check that you've updated the index by accessing the index of `aqi_sb`  by running `aqi_sb.index`. Add in a comment what is the data type (dtype) of the index.

In [ ]:
# solve exercise here

## 8
Run the next cell and read through the comments. They will explain how to do make an average over a rolling window.

- We can createCreate a new variable `five_day_average` with the mean of the AQI over a 5-day rolling window. To do this use the `rolling`
- 

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.rolling.html


In [ ]:
# rolling() is a method for pandas.series that provides rolling window calculations
# the parameter '5D' indicates we want the window to be 5 days
# This is a lazy method (think groupby), we need to specify what we want to calculate over each window
# here we add the aggregator function mean()
# this indicates we want the mean over each window
# and we get a pd.Series as ouput
aqi_sb.aqi.rolling('5D').mean()

## 9 

Without creating any new variables, add the mean of the AQI over a 5-day rolling window as a new column named 'five_day_average' to the `aqi_sb` dataframe. 

In [ ]:
# solve exercise here

## 10
Make a line plot showing both the daily AQI and the 5-day average (5-day average on top of the AQI). Update the title and colors of the graph. Can you see the AQI going up during the Thomas Fire in December 2017?

In [ ]:
# solve exercise here

## 11

Collect all the relevant code into the first blank cell of the notebook titled "FINAL CODE". This single cell will have the end-to-end workflow: from importing libraries and loading the data, to producing the graph. The *only* ouput of this cell should be the graph you produced in the previous exercise. For each line, add a single comment explaining what the code does.